### PR TITLE
Add parameters field to pass to a Jenkins job

### PR DESCRIPTION
**Pull request by katyafervent**
* New 'parameters' type is added
  It is allows to add parameter key and value and is saved as a dictionary
* New template and css style is added to support new type of field

### CODE REVIEW

File: kqueen_ui/blueprints/ui/templates/ui/partial/selectform.html [hunk 1]
```diff
@@ -134,4 +138,54 @@
     </span>
   {% endif %}
   </div>
+  {% endif %}
+{% endmacro %}
+
+{% macro render_multi_form_field(field) %}
+<div id="{{ field.name }}-fieldset"
```
Comment (ekhomyakova) [inline kqueen_ui/blueprints/ui/templates/ui/partial/selectform.html]: Could you please format as follows:
```
<div 
  id="{{ field.name }}-fieldset"
  class="form-group {% if field.errors %}has-error{% endif %} hidden"
  data-switchtag="{{ field.switchtag }}"
>
```
It's much more readable

File: kqueen_ui/blueprints/ui/templates/ui/partial/selectform.html [hunk 2]
```diff
@@ -134,4 +138,54 @@
...
+     class="form-group {% if field.errors %}has-error{% endif %} hidden"
+     data-switchtag="{{ field.switchtag }}">
+   {{ field.label }}
+   <button type="button"
+      class="btn btn-xs btn-default"
+      id="{{ field.name }}-fieldset-add-row"
+      data-toggle="tooltip"
+      title="Add new parameter"
+      >
+   +
```
Comment (ekhomyakova) [inline kqueen_ui/blueprints/ui/templates/ui/partial/selectform.html]: Move title here, like `<button ...>Add new parameter</button>`

File: kqueen_ui/blueprints/ui/templates/ui/partial/selectform.html [hunk 3]
```diff
@@ -134,4 +138,54 @@
...
+      {% for subform in field %}
+      <tr data-toggle="fieldset-entry">
+         {% for subfield in subform %}
+         <td> {{ subfield()|safe }} </td>
+         {% endfor %}
+         <td>
+            <button class="btn btn-xs btn-danger"
+               type="button"
+               data-toggle="tooltip"
+               title="Remove parameter"
```
Comment (ekhomyakova) [inline kqueen_ui/blueprints/ui/templates/ui/partial/selectform.html]: Move the title into the button content

File: kqueen_ui/blueprints/ui/views.py [hunk 4]
```diff
@@ -577,11 +577,11 @@ class ClusterCreate(KQueenView):
     def handle(self):
         # Get all necessary objects from backend
         _provisioners = self.kqueen_request('provisioner', 'list')
-        unknown_state = app.config['PROVISIONER_UNKNOWN_STATE']
-        ok_state = app.config['PROVISIONER_OK_STATE']
         provisioners = [
-            p for p in _provisioners
-            if p.get('state', unknown_state) == ok_state
+            p
```
Comment (ekhomyakova) [inline kqueen_ui/blueprints/ui/views.py]: Why do you need to add the line break? `p for p in _provisioners` fits the row perfectly

File: kqueen_ui/blueprints/ui/views.py [hunk 5]
```diff
@@ -594,7 +594,8 @@ def handle(self):
             # Append provisioner ID to parameter name to make it unique
             parameters = {
                 k + '__' + provisioner['id']: v
-                for [k, v] in _parameters.items()
+                for [k, v]
```
Comment (ekhomyakova) [inline kqueen_ui/blueprints/ui/views.py]: Why do you need to change it? 
It
is
less
readable
don't
you
think
so
?
)

File: kqueen_ui/blueprints/ui/views.py [hunk 6]
```diff
@@ -611,22 +612,23 @@ def handle(self):
                     in form._fields.items()
                     if (hasattr(v, 'switchtag') and v.switchtag) and form.provisioner.data in k
                 }
+                if 'override_parameters' in metadata.keys():
+                    metadata['override_parameters'] = {d['param_key']: d['param_value']
```
Comment (ekhomyakova) [inline kqueen_ui/blueprints/ui/views.py]:
```
metadata['override_parameters'] = {
    d['param_key']: d['param_value'] for d in metadata['override_parameters']
}
```

File: kqueen_ui/blueprints/ui/templates/ui/partial/selectform.html [hunk 7]
```diff
@@ -1,5 +1,6 @@
 {% macro render_selectform(form, action='') %}
-<form id="switchableForm" method="post" action="{{ action }}" enctype="multipart/form-data">
+ <script src="{{ url_for('static', filename='js/fieldlist.js') }}"></script>
```
Comment (ekhomyakova) [inline kqueen_ui/blueprints/ui/templates/ui/partial/selectform.html]: Did you add this file? Why do you need to import it?

File: kqueen_ui/asset/static/js/fieldlist.js [hunk 8]
```diff
@@ -0,0 +1,33 @@
+function fieldList() {
+    $("div[id$=fieldset]").each(function() {
```
Comment (ekhomyakova) [inline kqueen_ui/asset/static/js/fieldlist.js]: Could you please indent it with 2 spaces

File: kqueen_ui/asset/static/js/fieldlist.js [hunk 9]
```diff
@@ -0,0 +1,33 @@
+function fieldList() {
```
Comment (ekhomyakova) [inline kqueen_ui/asset/static/js/fieldlist.js]: I think this file should be in the `dynamic` folder

File: kqueen_ui/asset/static/js/fieldlist.js (hunk 8 above)
Comment (ekhomyakova) [inline kqueen_ui/asset/static/js/fieldlist.js]: Please use single quotes for JS

File: kqueen_ui/asset/static/js/fieldlist.js [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+    $("div[id$=fieldset]").each(function() {
+        var $this = $(this);
+        //Hide button that can remove last tr
```
Comment (ekhomyakova) [inline kqueen_ui/asset/static/js/fieldlist.js]: Please add a space after `//`

File: kqueen_ui/asset/static/js/fieldlist.js [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+        $(this).find("[data-toggle=fieldset-entry] td:last").hide();
```
Comment (ekhomyakova) [inline kqueen_ui/asset/static/js/fieldlist.js]: `$this`, since you've declared it

File: kqueen_ui/asset/static/js/fieldlist.js [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+
+        //Add new entry
+        $this.find("button[id$=fieldset-add-row]").click(function() {
+            var oldrow = $this.find("[data-toggle=fieldset-entry]:last");
+            var row = oldrow.clone(true, true);
+
+            var elem_id = row.find(":input")[0].id;
+            var elem_num = parseInt(elem_id.replace(/.*-(\d{1,4})-.*/m, '$1')) + 1;
```
Comment (ekhomyakova) [inline kqueen_ui/asset/static/js/fieldlist.js]: Please add a comment, it's not so clear logic

File: kqueen_ui/asset/static/js/fieldlist.js [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+            row.attr('data-id', elem_num);
+            row.find(":input").each(function() {
+
+                var id = $(this).attr('id').replace('-' + (elem_num - 1) + '-', '-' + (elem_num) + '-');
+                $(this).attr('name', id).attr('id', id).val('');
+            });
+            row.find('td:last').show()
```
Comment (ekhomyakova) [inline kqueen_ui/asset/static/js/fieldlist.js]: Please add `;` in the end

File: kqueen_ui/asset/static/js/fieldlist.js [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+        var $this = $(this);
+        //Hide button that can remove last tr
+        $(this).find("[data-toggle=fieldset-entry] td:last").hide();
+
+        //Add new entry
+        $this.find("button[id$=fieldset-add-row]").click(function() {
+            var oldrow = $this.find("[data-toggle=fieldset-entry]:last");
+            var row = oldrow.clone(true, true);
+
+            var elem_id = row.find(":input")[0].id;
```
Comment (ekhomyakova) [inline kqueen_ui/asset/static/js/fieldlist.js]: And JS is about camelCase

File: kqueen_ui/utils/fields.py [hunk 15]
```diff
@@ -115,3 +117,15 @@ def process_formdata(self, valuelist):
                 msg = 'Provided file is not a YAML'
                 logger.exception(msg)
                 raise ValueError(msg)
+
+
+class ParamForm(WTForm):
+    param_key = StringField('Parameter', render_kw={"placeholder": "Existent parameter name"})
```
Comment (ekhomyakova) [inline kqueen_ui/utils/fields.py]: I suggest
```
param_key = StringField('Existent parameter name')
```

File: kqueen_ui/asset/dynamic/scss/main.scss [hunk 16]
```diff
@@ -370,6 +370,17 @@ label.error {
   box-shadow: inset 0px 0px 0px 0px white;
 }
 
+.fieldset-table {
+  margin: 15px;
+}
+
+.fieldset-table td {
```
Comment (ekhomyakova) [inline kqueen_ui/asset/dynamic/scss/main.scss]: You can group it under `.fieldset-table {}`

File: kqueen_ui/asset/static/js/fieldlist.js [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+
+        var id = $(this).attr('id').replace('-' + (elemNum - 1) + '-', '-' + (elemNum) + '-');
+        $(this).attr('name', id).attr('id', id).val('');
+      });
+      row.find('td:last').show();
+      oldRow.after(row);
+    });
+
+    //  Remove row
+    $this.find('button[id$==remove-row]').click(function () {
```
Comment (ekhomyakova) [inline kqueen_ui/asset/static/js/fieldlist.js]: What does this selector (`$==`) do?

File: kqueen_ui/blueprints/ui/templates/ui/partial/selectform.html [hunk 18]
```diff
@@ -134,4 +138,56 @@
     </span>
   {% endif %}
   </div>
+  {% endif %}
+{% endmacro %}
+
+{% macro render_multi_form_field(field) %}
+    <div
```
Comment (ekhomyakova) [inline kqueen_ui/blueprints/ui/templates/ui/partial/selectform.html]: As I noticed, we use 2 indentation spaces for html

File: kqueen_ui/blueprints/ui/templates/ui/partial/selectform.html [hunk 19]
```diff
@@ -134,4 +138,56 @@
...
+      id="{{ field.name }}-fieldset"
+      class="form-group {% if field.errors %}has-error{% endif %} hidden"
+      data-switchtag="{{ field.switchtag }}"
+    >
+        {{ field.label }}
+        <button type="button"
```
Comment (ekhomyakova) [inline kqueen_ui/blueprints/ui/templates/ui/partial/selectform.html]: Could you please format this as a div above?

File: kqueen_ui/blueprints/ui/templates/ui/partial/selectform.html [hunk 20]
```diff
@@ -134,4 +138,56 @@
...
+                                id="{{ field.name }}-{{ loop.index0 }}-remove-row">
+                            -
+                        </button>
+                    </td>
+                </tr>
+            {% endfor %}
+        </table>
+        {% if field.errors %}
+            <span class="help-block">
+    {% for error in field.errors %}
```
Comment (ekhomyakova) [inline kqueen_ui/blueprints/ui/templates/ui/partial/selectform.html]: Looks like smth is wrong with indentation here

File: kqueen_ui/blueprints/ui/templates/ui/partial/selectform.html [hunk 21]
```diff
@@ -141,4 +145,58 @@
     </span>
   {% endif %}
   </div>
+  {% endif %}
+{% endmacro %}
+
+{% macro render_multi_form_field(field) %}
+  <div
+      id="{{ field.name }}-fieldset"
```
[the remark this answers is not in the document]
Comment (katyafervent) [reply kqueen_ui/blueprints/ui/templates/ui/partial/selectform.html]: where this rule is defined?

File: kqueen_ui/blueprints/ui/templates/ui/partial/selectform.html [hunk 22]
```diff
@@ -118,27 +119,84 @@
...
+          {% for subfield in subform %}
+            <td> {{ subfield(class="form-control")|safe }} </td>
+          {% endfor %}
+          <td>
+            <button
+              class="btn btn-xs btn-danger"
+              type="button"
+              data-toggle="tooltip"
+              title="Remove parameter"
+              id="{{ field.name }}-{{ loop.index0 }}-remove-row">
```
Comment (ekhomyakova) [inline kqueen_ui/blueprints/ui/templates/ui/partial/selectform.html]: `>` should be in the new line...

**Pull request merged**